From cb66c88ad871a4d9c989ba240db749311a25c9de Mon Sep 17 00:00:00 2001
From: Jiang Lu <lu.jiang@windriver.com>
Date: Thu, 28 Jun 2018 16:29:45 +0800
Subject: [PATCH] ostree-sysroot-deploy: add OSTREE_ETC_MERGE_DIR env var

If OSTREE_ETC_MERGE_DIR = "none" skip merge.

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>

Upstream-Status: Inappropriate [WRLinux specific]

This env var, OSTREE_ETC_MERGE_DIR, is used by ostree_upgrade.sh
which is WRLinux specific. The default value is '/etc', but has
option to set it to 'none'. We'll need to keep this patch until
we change ostree_upgrade.sh.

Signef-off-by: Chen Qi <Qi.Chen@windriver.com>
---
 src/libostree/ostree-sysroot-deploy.c | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/src/libostree/ostree-sysroot-deploy.c b/src/libostree/ostree-sysroot-deploy.c
index f7ca2dd4..960c52fd 100644
--- a/src/libostree/ostree-sysroot-deploy.c
+++ b/src/libostree/ostree-sysroot-deploy.c
@@ -524,7 +524,16 @@ merge_configuration_from (OstreeSysroot *sysroot, OstreeDeployment *merge_deploy
 
   /* TODO: get rid of GFile usage here */
   g_autoptr (GFile) orig_etc = ot_fdrel_to_gfile (merge_deployment_dfd, "usr/etc");
-  g_autoptr (GFile) modified_etc = ot_fdrel_to_gfile (merge_deployment_dfd, "etc");
+  char *etc_merge_dir = getenv("OSTREE_ETC_MERGE_DIR");
+  g_autoptr(GFile) modified_etc;
+  if (etc_merge_dir) {
+    /* if OSTREE_ETC_MERGE_DIR = "none" skip merge */
+    if (!strcmp(etc_merge_dir, "none"))
+      return TRUE;
+    modified_etc = g_file_new_for_path(etc_merge_dir);
+  } else {
+    modified_etc = ot_fdrel_to_gfile (merge_deployment_dfd, "etc");
+  }
   /* Return values for below */
   g_autoptr (GPtrArray) modified
       = g_ptr_array_new_with_free_func ((GDestroyNotify)ostree_diff_item_unref);
@@ -556,8 +565,15 @@ merge_configuration_from (OstreeSysroot *sysroot, OstreeDeployment *merge_deploy
   if (!glnx_opendirat (merge_deployment_dfd, "usr/etc", TRUE, &orig_etc_fd, error))
     return FALSE;
   glnx_autofd int modified_etc_fd = -1;
-  if (!glnx_opendirat (merge_deployment_dfd, "etc", TRUE, &modified_etc_fd, error))
-    return FALSE;
+  if (etc_merge_dir)
+    {
+      if (!glnx_opendirat (-1, etc_merge_dir, TRUE, &modified_etc_fd, error))
+        return FALSE;
+    }
+  else
+    if (!glnx_opendirat (merge_deployment_dfd, "etc", TRUE, &modified_etc_fd, error))
+      return FALSE;
+
   glnx_autofd int new_etc_fd = -1;
   if (!glnx_opendirat (new_deployment_dfd, "etc", TRUE, &new_etc_fd, error))
     return FALSE;
-- 
2.45.1

